### PR TITLE
(extension) deduplicate danmaku with opt-out whitelist [DA-457]

### PR DESCRIPTION
**Pull request by Mr-Quin**
## Summary
- Add `dedupComments()` engine utility that hides exact text duplicates within a configurable timestamp tolerance window
- Whitelist (regex/text) exempts common reaction danmaku (哈哈, www, 666, etc.) from deduplication
- New "Dedup" tab in the Filter page with enable toggle, tolerance slider, and whitelist management
- CommentsTable gains a local "Hide duplicates" toggle
- Settings sync + backup automatically via existing DanmakuOptions infrastructure
- Migration v8 seeds defaults for existing installs

## Test plan
- [ ] Engine unit tests for `dedupComments` cover all edge cases (tolerance, whitelist, stability, disabled)
- [ ] Migration test for v8
- [ ] Manual: open dev browser, load danmaku with known duplicates, verify duplicates hidden
- [ ] Manual: toggle dedup off in Filter > Dedup tab, verify duplicates reappear
- [ ] Manual: add whitelist entry, verify matching duplicates are preserved
- [ ] Manual: CommentsTable toggle hides/shows duplicates independently of global setting

### CODE REVIEW

**Comment by Mr-Quin** [comment]
Task linked: [DA-457 Deduplicate danmaku with opt-out whitelist](https://app.clickup.com/t/90131020449/DA-457)

File: packages/danmaku-anywhere/src/common/components/CommentsTable.tsx [hunk 1]
```diff
@@ -69,13 +71,28 @@ export const CommentsTable = ({
   const [hoverRow, setHoverRow] = useState<number>()
   const [filter, setFilter] = useState('')
 
+  const { data: danmakuOptions } = useDanmakuOptions()
+  const [hideDuplicates, setHideDuplicates] = useState(
+    danmakuOptions.dedup.enabled
+  )
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-anywhere/src/common/components/CommentsTable.tsx]: Fixed — now uses `useState<boolean | null>(null)` with fallback to global setting, so it stays in sync until the user explicitly toggles.

File: packages/danmaku-anywhere/src/common/components/DanmakuFilter/FilterPageCore.tsx [hunk 2]
```diff
@@ -74,35 +89,165 @@ export const FilterPageCore = ({
...
+          <Typography variant="subtitle2" gutterBottom>
+            {t('danmakuFilter.dedup.toleranceLabel')}
+          </Typography>
+          <Slider
+            value={config.dedup.tolerance}
+            onChange={(_, value) => {
+              handleUpdate((draft) => {
+                draft.dedup.tolerance = value as number
+              })
+            }}
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-anywhere/src/common/components/DanmakuFilter/FilterPageCore.tsx:201]: Fixed — slider now uses local state with \ to persist only when the user finishes dragging.

File: packages/danmaku-engine/src/parser.ts [hunk 3]
```diff
@@ -101,3 +101,51 @@ export const filterComments = (
...
+
+  // Create index pairs so we can sort by time but restore to mark kept/dropped
+  const indexed = comments.map((c, i) => ({
+    comment: c,
+    time: parseCommentEntityP(c.p).time,
+    originalIndex: i,
+  }))
+
+  // Stable sort by time
+  indexed.sort((a, b) => a.time - b.time)
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-engine/src/parser.ts]: Fixed — added \ tie-breaker to the sort comparator for deterministic ordering.

File: packages/danmaku-engine/src/parser.ts [hunk 4]
```diff
@@ -101,3 +101,51 @@ export const filterComments = (
...
+
+  // Track which original indices are kept
+  const kept = new Set<number>()
+  const lastKeptTime = new Map<string, number>()
+
+  for (const { comment, time, originalIndex } of indexed) {
+    // Check if whitelisted (exempt from dedup)
+    if (applyFilter(comment.m, config.whitelist)) {
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-engine/src/parser.ts]: Fixed — added a per-message \ so each unique text only runs \ once per \ call.

File: packages/danmaku-anywhere/src/common/options/danmakuOptions/constant.ts [hunk 5]
```diff
@@ -1,10 +1,22 @@
...
+  DanmakuOptions as DanmakuEngineOptions,
+  DanmakuFilter,
+} from '@danmaku-anywhere/danmaku-engine'
 
 import type { Options } from '@/common/options/OptionsService/types'
 
-export type DanmakuOptions = Omit<DanmakuEngineOptions, 'show'> & {
+export interface DedupOptions {
+  readonly enabled: boolean
+  /** Seconds tolerance for same-timestamp bucket. */
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-anywhere/src/common/options/danmakuOptions/constant.ts]: Fixed — updated the JSDoc to describe the rolling +/- time window semantics.

File: packages/danmaku-anywhere/src/common/components/DanmakuFilter/FilterPageCore.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-anywhere/src/common/components/DanmakuFilter/FilterPageCore.tsx:201]: Fixed — same as the Gemini comment above. Slider now uses \.

File: packages/danmaku-engine/src/parser.ts [hunk 6]
```diff
@@ -101,3 +101,63 @@ export const filterComments = (
...
+  const lastKeptTime = new Map<string, number>()
+
+  // Cache whitelist results per message text to avoid recompiling regexes
+  const whitelistCache = new Map<string, boolean>()
+  const isWhitelisted = (text: string): boolean => {
+    const cached = whitelistCache.get(text)
+    if (cached !== undefined) {
+      return cached
+    }
+    const matched = applyFilter(text, config.whitelist)
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-engine/src/parser.ts:138]: Already addressed — the per-message \ added in the previous fix means each unique message text only calls \ (and its regex compilation) once. Since duplicates share the same text, the cache is very effective here.

File: packages/danmaku-anywhere/src/common/options/danmakuOptions/constant.ts [hunk 7]
```diff
@@ -1,10 +1,22 @@
...
+  readonly tolerance: number
+  /** Comments whose text matches any enabled entry are exempted from dedup. */
+  readonly whitelist: DanmakuFilter[]
+}
+
+export type DanmakuOptions = Omit<DanmakuEngineOptions, 'show' | 'dedup'> & {
   readonly customCss: string
   readonly useCustomCss: boolean
+  readonly dedup: DedupOptions
 }
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-anywhere/src/common/options/danmakuOptions/constant.ts:20]: Intentional — \ uses \ as part of the extension's immutable options pattern (consistent with all other option types in this layer). TypeScript structural typing handles readonly→mutable assignment fine. The two types live in different packages serving different layers (engine config vs extension storage).

File: packages/danmaku-anywhere/src/common/components/DanmakuFilter/FilterPageCore.tsx [hunk 8]
```diff
@@ -74,35 +90,169 @@ export const FilterPageCore = ({
...
+        draft.dedup.whitelist.push({
+          type: 'text',
+          value: result.pattern,
+          enabled: true,
+        })
+      })
+    }
+    setDedupFilterError('')
+    return true
+  }
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-anywhere/src/common/components/DanmakuFilter/FilterPageCore.tsx:125]: Agreed the duplication exists, but this mirrors the pre-existing pattern in \ — refactoring the shared validation+update logic is out of scope for this PR.

File: packages/danmaku-engine/src/parser.ts [hunk 9]
```diff
@@ -101,3 +101,63 @@ export const filterComments = (
...
+    whitelistCache.set(text, matched)
+    return matched
+  }
```
[the remark this answers is not in the document]
Comment (Mr-Quin) [reply packages/danmaku-engine/src/parser.ts:141]: Already addressed — the per-message \ means regex compilation only happens once per unique message text, which is sufficient for typical comment set sizes.
